Add --chunk-size flag to kubectl drain

KnVerey opened the pull request:
#### What type of PR is this?

/kind feature


#### What this PR does / why we need it:

Adds the `--chunk-size` flag to kubectl drain . This flag causes the `limit` parameter to be set on list requests the command makes, which in practice is a node list and a pod list. The flag's value defaults to 500, the same as `kubectl get`.

Having this flag available with a sane default value improves the scalability of the command because list requests made without a limit parameter (as `drain` is sending today) can cause a full load of the underlying resource data from etcd to be attempted, and this may fail on clusters of sufficient scale. For example, this change may help you if you are seeing grpc message size errors when attempting to drain in a large cluster, even when you're targeting a single node running a small number of pods. For more detail on the problem and solutions considered, please refer to this issue: https://github.com/kubernetes/kubectl/issues/1028.

#### Which issue(s) this PR fixes:

Part 1 of fixing https://github.com/kubernetes/kubectl/issues/1028

#### Special notes for your reviewer:

I initially used the Selector Visitor in an attempt to follow established patterns; this is where `continue` support was added when chunking was added to `get`, and it does work for this case too. However, in a later commit I did some refactoring to extract the `continue` handling, making it easier to add chunking support to places like `drain` that are using a specific client right now.


#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
`kubectl drain` will by default fetch large lists of resources in chunks of up to 500 items rather than requesting all resources up front from the server. A new flag `--chunk-size=SIZE` may be used to alter the number of items or disable this feature when `0` is passed.  This is a beta feature.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

The flag is included in kubectl's own flag documentation, i.e.:

```
Options:
      --chunk-size=500: Return large lists in chunks rather than all at once. Pass 0 to disable. This flag is beta and may change in the future.
```

If other documentation updates need to be made, please let me know.

KnVerey committed:
Extract DefaultChunkSize constant

eddiezane commented on staging/src/k8s.io/kubectl/pkg/cmd/drain/drain.go:
```diff
@@ -197,6 +198,8 @@ func NewCmdDrain(f cmdutil.Factory, ioStreams genericclioptions.IOStreams) *cobr
 	cmd.Flags().StringVarP(&o.drainer.PodSelector, "pod-selector", "", o.drainer.PodSelector, "Label selector to filter pods on the node")
 	cmd.Flags().BoolVar(&o.drainer.DisableEviction, "disable-eviction", o.drainer.DisableEviction, "Force drain to use delete, even if eviction is supported. This will bypass checking PodDisruptionBudgets, use with caution.")
 	cmd.Flags().IntVar(&o.drainer.SkipWaitForDeleteTimeoutSeconds, "skip-wait-for-delete-timeout", o.drainer.SkipWaitForDeleteTimeoutSeconds, "If pod DeletionTimestamp older than N seconds, skip waiting for the pod.  Seconds must be greater than 0 to skip.")
+	cmd.Flags().Int64Var(&o.drainer.ChunkSize, "chunk-size", o.drainer.ChunkSize,
```
Let's use a helper here (hopping to clean up all these duplicate strings one day...).

https://github.com/kubernetes/kubernetes/blob/60387551078a93baddea8b1012ff0adedc10cae2/staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go#L437

KnVerey commented:
/test pull-kubernetes-e2e-kind

KnVerey commented on test/cmd/node-management.sh:
```diff
@@ -35,10 +35,12 @@ run_cluster_management_tests() {
     }
   },
   "spec": {
+  "nodeName": "127.0.0.1",
```
The `drain` tests in this file are fundamentally broken because of the lack of this label: they are draining a node with no pods assigned to it, which of course succeeds, but rather meaninglessly.

KnVerey commented on test/cmd/node-management.sh:
```diff
@@ -109,8 +113,10 @@ __EOF__
+  # command - need to use force because pods are unmanaged
+  kubectl drain "127.0.0.1" --dry-run=client --force
+  # TODO: reinstate this when server-side dry run is fixed. See https://github.com/kubernetes/kubernetes/pull/100206
+  # kubectl drain "127.0.0.1" --dry-run=server --force
```
As noted in https://github.com/kubernetes/kubernetes/pull/100206#issue-592252171, this command fails with production builds of kubectl. This suite isn't catching it on master because of the aforementioned lack of pods on the target node. Error is:

```
error: error when evicting pods/"$PODNAME" -n "$NAMESPACE": /, Kind= doesn't support dry-run
```

KnVerey commented on test/cmd/node-management.sh:
```diff
@@ -120,16 +126,16 @@ __EOF__
-  # command
-  kubectl drain "127.0.0.1" --pod-selector 'e in (f)'
+  # command - need to use force because pods are unmanaged
+  # TODO: reinstate this when it works on CI. See https://github.com/kubernetes/kubernetes/pull/100206
```
As mentioned in https://github.com/kubernetes/kubernetes/pull/100206, when I fix this suite so that the test pods are actually on the target node, the drain command seems to hang forever in CI specifically, and I haven't figured out why. 

We can block this PR on the other one if you'd like, but please note that this assertion isn't proving all that much right now, with no pods on the node.

pwittrock commented on staging/src/k8s.io/cli-runtime/pkg/resource/helper.go:
```diff
@@ -90,6 +93,49 @@ func (m *Helper) List(namespace, apiVersion string, options *metav1.ListOptions)
 	return req.Do(context.TODO()).Get()
 }
 
+func FollowContinue(initialOpts *metav1.ListOptions,
```
Document public functions

pwittrock commented:
/approve

pwittrock commented:
/triage accepted

pwittrock commented:
/hold

Hold on the tests getting resolved

pwittrock commented:
/lgtm
/approve

note: can wait until end of code-freeze

KnVerey committed:
Create helper for adding chunk-size flag to commands

KnVerey committed:
Extract helper for continue handling

KnVerey committed:
Add --chunk-size flag to kubectl drain

KnVerey commented:
Rewrote the commits as requested and added documentation to the new helpers. This PR now includes the commits from https://github.com/kubernetes/kubernetes/pull/100206 required for CI to pass without disabling assertions. When code freeze is over, we should merge https://github.com/kubernetes/kubernetes/pull/100206 first and then rebase this one.

KnVerey commented:
/retest

KnVerey commented:
/retest

pwittrock commented:
/lgtm

KnVerey commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd
/test pull-kubernetes-e2e-kind

sftim commented:
/retest

KnVerey commented:
/test pull-kubernetes-integration

pwittrock commented:
/lgtm

pwittrock commented:
/hold cancel
